components/pd_client/src/tso.rs

```diff
@@ -0,0 +1,236 @@
+            if pending_requests.len() == MAX_PENDING_COUNT {
+                sending_future_waker.wake();
+            }
+
+            allocate_timestamps(&resp, &mut pending_requests)?;
+        }
+        Ok(())
+    };
+
+    let (send_res, recv_res): (Result<()>, Result<()>) =
+        join!(send_requests, receive_and_handle_responses);
+    let _ = close_tx.send(());
+    info!("TSO worker terminated"; "sender_cause" => ?send_res.err(), "receiver_cause" => ?recv_res.err());
+}
+
+struct RequestGroup {
+    tso_request: TsoRequest,
+    requests: Vec<TimestampRequest>,
+}
+
+struct TsoRequestStream<'a> {
+    cluster_id: u64,
+    request_rx: &'a mut mpsc::Receiver<oneshot::Sender<TimeStamp>>,
+    pending_requests: Rc<RefCell<VecDeque<RequestGroup>>>,
+    self_waker: Rc<AtomicWaker>,
+}
+
+impl<'a> Stream for TsoRequestStream<'a> {
+    type Item = (TsoRequest, WriteFlags);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let pending_requests = self.pending_requests.clone();
+        let mut pending_requests = pending_requests.borrow_mut();
+        let mut requests = Vec::new();
+
+        while requests.len() < MAX_BATCH_SIZE && pending_requests.len() < MAX_PENDING_COUNT {
+            match self.request_rx.poll_recv(cx) {
+                Poll::Ready(Some(sender)) => {
+                    requests.push(sender);
+                }
+                Poll::Ready(None) if requests.is_empty() => {
+                    return Poll::Ready(None);
+                }
+                _ => break,
+            }
+        }
+
+        if !requests.is_empty() {
+            let mut req = TsoRequest::default();
+            req.mut_header().cluster_id = self.cluster_id;
+            req.count = requests.len() as u32;
+
+            let request_group = RequestGroup {
+                tso_request: req.clone(),
+                requests,
+            };
+            pending_requests.push_back(request_group);
+
+            let write_flags = WriteFlags::default().buffer_hint(false);
+            Poll::Ready(Some((req, write_flags)))
+        } else {
+            // Set the waker to the context, then the stream can be waked up after the pending queue
+            // is no longer full.
+            self.self_waker.register(cx.waker());
+            Poll::Pending
+        }
+    }
+}
+
+fn allocate_timestamps(
+    resp: &TsoResponse,
+    pending_requests: &mut VecDeque<RequestGroup>,
+) -> Result<()> {
+    // PD returns the timestamp with the biggest logical value. We can send back timestamps
+    // whose logical value is from `logical - count + 1` to `logical` using the senders
+    // in `pending`.
+    let tail_ts = resp
+        .timestamp
+        .as_ref()
+        .ok_or_else(|| -> Error { box_err!("No timestamp in TsoResponse") })?;
+
+    let mut offset = resp.count;
+    if let Some(RequestGroup {
+        tso_request,
+        requests,
+    }) = pending_requests.pop_front()
```

Comment:
Can we guarantee the order of `resp` matches the `pending_requests`?

Reply:
I believe PD handles requests in a single stream RPC serially. So the order should be the same.